pkg/metrics/collector/kernel_metrics.go

```diff
@@ -0,0 +1,145 @@
+const (
+	// Got from Snmp counters
+	TCPSegsIn      = "InSegs"      // Total tcp segments in
+	TCPSegsOut     = "OutSegs"     // Total tcp segments out
+	TCPSegsRetrans = "RetransSegs" // Total tcp segments retransmitted
+	// Got from netstat
+	TCPTimeoutRehash   = "TcpTimeoutRehash"      // Total tcp timeout rehash
+	TCPDuplicateRehash = "TcpDupliateDataRehash" // Total tcp duplicates rehashed
+)
+
+var (
+	tcpTimeoutDesc = prometheus.NewDesc(
+		"tcp_timeout_rehash_count",
+		"Tcp timeout rehash count",
+		nil, nil,
+	)
+	tcpDuplicateDesc = prometheus.NewDesc(
+		"tcp_duplicate_rehash_count",
+		"Tcp duplicate rehash counts",
+		nil, nil,
+	)
+	segmentsReceivedDesc = prometheus.NewDesc(
+		"tcp_segments_received_count",
+		"TCP segments received on the node",
+		nil, nil,
+	)
+	segmentsSentDesc = prometheus.NewDesc(
+		"tcp_segments_sent_count",
+		"TCP segments sent from node",
+		nil, nil,
+	)
+	segmentsRetransmittedDesc = prometheus.NewDesc(
+		"tcp_segments_retransmitted_count",
+		"TCP segments retransmitted on the node",
+		nil, nil,
+	)
+)
+
+type kernelStatCollector struct {
+}
+
+func init() {
+	registerCollector("kernel_metrics", NewKernelStatCollector)
+}
+
+func NewKernelStatCollector() (Collector, error) {
+	return &kernelStatCollector{}, nil
+}
+
+// Parses a string which contains two lines, each starting with label
+// returng a map where the keys are the first lines elements, while the
+// second line contains the values, all casted to uint64, e.g
+//
+// label key1 key2 key3
+// label 1 2 3
+//
+// Outputs: {"key1":1 "key2":2 "key3":3}
+func parseKeyValueLines(output, label string) map[string]uint64 {
+	stats := make(map[string]uint64)
+	splitFile := strings.Split(output, "\n")
+	for i, stat := range splitFile {
+
+		if !strings.Contains(stat, label) {
+			continue
+		}
+		keyList := strings.Fields(stat)
+		// TODO: Look for next instance of label rather than assuming its the next
+		valList := strings.Fields(splitFile[i+1])
+		for i, key := range keyList[1:] {
+			val, err := strconv.Atoi(valList[i+1])
+			if err != nil {
+				continue
+			}
+			stats[key] = uint64(val)
+		}
+		return stats
+	}
+	return stats
+}
+
+func (c *kernelStatCollector) Update(ch chan<- prometheus.Metric) error {
+	// Get snmp values.
+	data, err := ioutil.ReadFile("/proc/net/snmp")
+	if err == nil {
+		output := string(data)
+		netstats := parseKeyValueLines(output, SnmpLabel)
+		if val, ok := netstats[TCPSegsIn]; ok {
+			ch <- prometheus.MustNewConstMetric(segmentsReceivedDesc, prometheus.CounterValue, float64(val))
+		}
+		if val, ok := netstats[TCPSegsOut]; ok {
+			ch <- prometheus.MustNewConstMetric(segmentsSentDesc, prometheus.CounterValue, float64(val))
+		}
+		if val, ok := netstats[TCPSegsRetrans]; ok {
+			ch <- prometheus.MustNewConstMetric(segmentsRetransmittedDesc, prometheus.CounterValue, float64(val))
+		}
+
+	}
+
+	// Get netstat values
+	data, err = ioutil.ReadFile("/proc/net/netstat")
+	if err != nil {
+		return nil
```

Comment:
Here and the above - let's return the error?

Reply:
I do see in `conntrack.go` we ignore a similar error and return `nil` but I don't think we should hide it :/ I will fix that separately.

Reply:
Sounds good!